File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/remoteconfiguration/FetchedConfigurationBundle.java
package com.snowplowanalytics.snowplow.internal.remoteconfiguration;

import android.content.Context;

import androidx.annotation.NonNull;

import com.snowplowanalytics.snowplow.configuration.Configuration;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class FetchedConfigurationBundle implements Configuration {

    @NonNull
    public String schema;

    public int configurationVersion;

    @NonNull
    public List<ConfigurationBundle> configurationBundle;

    FetchedConfigurationBundle(@NonNull String schema) {
        this.schema = schema;
        this.configurationVersion = -1;
        this.configurationBundle = new ArrayList<>();
    }

    // JSON formatter

    public FetchedConfigurationBundle(@NonNull Context context, @NonNull JSONObject jsonObject) throws JSONException {
        schema = jsonObject.getString("$schema");
        configurationVersion = jsonObject.getInt("configurationVersion");
        configurationBundle = new ArrayList<>();
        JSONArray array = jsonObject.getJSONArray("configurationBundle");
        for (int i = 0; i < array.length(); i++) {
            JSONObject bundleJson = array.getJSONObject(i);
            ConfigurationBundle bundle = new ConfigurationBundle(context, bundleJson);
            configurationBundle.add(bundle);
        }
    }

    // Copyable

    @NonNull
    @Override
    public Configuration copy() {
        FetchedConfigurationBundle copy = new FetchedConfigurationBundle(schema);
        copy.configurationVersion = configurationVersion;
        for (ConfigurationBundle bundle : configurationBundle) {
            copy.configurationBundle.add((ConfigurationBundle)bundle.copy());
        }
        return copy;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/remoteconfiguration/FetchedConfigurationBundle.kt
package com.snowplowanalytics.core.remoteconfiguration

import android.content.Context
import com.snowplowanalytics.snowplow.configuration.Configuration
import com.snowplowanalytics.snowplow.configuration.ConfigurationBundle
import org.json.JSONObject

class FetchedConfigurationBundle : Configuration {
    var schema: String
    var configurationVersion: Int
    var configurationBundle: List<ConfigurationBundle>

    constructor(schema: String) {
        this.schema = schema
        configurationVersion = -1
        configurationBundle = listOf()
    }

    // JSON formatter
    constructor(context: Context, jsonObject: JSONObject) {
        schema = jsonObject.getString("\$schema")
        configurationVersion = jsonObject.getInt("configurationVersion")
        val tempBundle = ArrayList<ConfigurationBundle>()
        
        val array = jsonObject.getJSONArray("configurationBundle")
        for (i in 0 until array.length()) {
            val bundleJson = array.getJSONObject(i)
            val bundle = ConfigurationBundle(context, bundleJson)
            tempBundle.add(bundle)
        }
        configurationBundle = tempBundle.toList()
    }

    // Copyable
    override fun copy(): Configuration {
        val copy = FetchedConfigurationBundle(schema)
        copy.configurationVersion = configurationVersion
        
        val tempBundle = ArrayList<ConfigurationBundle>()
        for (bundle in configurationBundle) {
            (bundle.copy() as? ConfigurationBundle)?.let { tempBundle.add(it) }
        }
        copy.configurationBundle = tempBundle.toList()
        return copy
    }
}
